Add recall to dense vector track

jdconrad opened the pull request:
Recall in dense vectors is a measurement of accuracy between an approximate knn search and a brute force script score. Accuracy can be measured by figuring out the count of the intersection between the documents returned by both queries and divided by the size of the script score query. k should match size for this to be meaningful.

This change adds a new track runner (knn-recall) and a new params source (knn-recall-params-source) to support collecting new data added to the meta field in the form of `k`, `num-candidates`, and `recall` which can be used to determine how accurate a specific query was.

The source document generated for the metrics store entries for this runner look like the following:

```
"_source" : {
          "@timestamp" : 1670429366132,
          "relative-time" : 829.0115890558809,
          "race-id" : "85da22bb-762b-4c5c-9f07-b7760d8b9245",
          "race-timestamp" : "20221207T153222Z",
          "environment" : "local",
          "track" : "dense_vector",
          "challenge" : "index-and-search",
          "car" : "defaults",
          "name" : "latency",
          "value" : 246.23832199722528,
          "unit" : "ms",
          "sample-type" : "normal",
          "meta" : {
            "attribute_xpack.installed" : "true",
            "source_revision" : "a334a530efa1460872ebb2bf53706cba3fbdac9e",
            "distribution_version" : "8.7.0-SNAPSHOT",
            "distribution_flavor" : "default",
            "k" : 10,
            "num_candidates" : 100,
            "recall" : 10,
            "success" : true
          },
          "task" : "knn-recall-10-100",
          "operation" : "knn-recall-10-100",
          "operation-type" : "knn-recall"
        }
```

jdconrad commented on dense_vector/challenges/default.json:
```diff
@@ -126,6 +126,41 @@
+      "operation": "knn-recall-10-100",
+      "warmup-iterations": 0,
+      "iterations": 1000,
+      "clients": 1
```
I see, I will remove these.

jdconrad commented on dense_vector/challenges/default.json:
```diff
@@ -126,6 +126,41 @@
+    {
+      "name": "knn-recall-10-100",
+      "operation": "knn-recall-10-100",
+      "warmup-iterations": 0,
+      "iterations": 1000,
```
Sure, let me give that a try. My only concern is maintaining enough information to debug appropriately, but I guess if we see a drop in recall that's concerning, it's not a difficult change to figure out which queries are worse off locally.

benwtrent replied:
> it's not a difficult change to figure out which queries are worse off locally.

This is what we have to do with rally now. If there is a significant change and we cannot determine it via look at the change log, then we try to reproduce it :)

jdconrad commented:
@DJRickyB Thanks for the first review. This is ready for the next one. I did the following:
1. Updated operations to set `include-in-reporting` to `false`
2. Moved all queries to be executed in a single pass of the runner and only return a single result for recall now. This seems to only generate 1 record per shard from what I can tell locally.
3. Removed the extraneous values from the `*-recall-*` challenges.
4. Ran pre-commit to ensure the style is compliant.

benwtrent commented on dense_vector/track.py:
```diff
@@ -72,69 +74,73 @@ def __init__(self, track, params, **kwargs):
-            "_source": False,
+        return {
+            "index": self._index_name,
+            "cache": self._params.get("cache", False),
```
Why are we defaulting to `cache` being false? The underlying dataset isn't changing, so caching the result set is perfectly fine IMO. Is there something I am missing?

jdconrad replied:
I pulled these settings from the latency/throughput runs for knn search where I assume caching could change our performance test results in an undesirable way. For this test, we don't use the same query vector twice (or general query), so wouldn't caching potentially use additional resources that end up being wasted? It's certainly possible my understanding isn't correct here, though.

jdconrad commented on dense_vector/track.py:
```diff
@@ -56,5 +56,91 @@ def params(self):
+            intersection += len(knn_hits.intersection(script_hits))
+            total += len(script_hits)
+
+        return {"recall": intersection / total}
```
I will change this to avg_recall -- this is just better naming even if we don't extend this to have more parameters.

I think it's probably fine to abort this test on empty results, but I will guard for the 0/0 case. Thanks for pointing this out.

jdconrad replied:
I'm also adding min_recall now since this does seem useful to identify if there are any really bad queries. Thank you for the suggestion.

jdconrad commented:
@DJRickyB I update this change to include the following:

1. renamed `recall` to `avg_recall`
2. added `min_recall` to the meta dict
3. protected against a possible divide by 0 error (in theory this should never actually happen)

benwtrent reviewed (approved):
:shipit: 

Love this!

jdconrad commented:
@DJRickyB and @benwtrent Thank you both for the reviews!
